src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,210 @@
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
+            if (folderItems != null)
+            {
+                Parallel.ForEach(folderItems, (folder) =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, folder, artifact.Name);
+                    Directory.CreateDirectory(targetPath);
+                });
+            }
+
+            var fileItems = items.Where(i => i.ItemType == ContainerItemType.File);
+
+            var batchItemsBlock = new BatchBlock<FileContainerItem>(
+                batchSize: 1000,
+                new GroupingDataflowBlockOptions()
+                {
+                    CancellationToken = cancellationToken,
+                });
+
+            var fetchStream = NonSwallowingTransformManyBlock.Create<IEnumerable<FileContainerItem>, (Stream, string)>(
+                async itemsStream =>
+                {
+                    List<(Stream, string)> collection= new List<(Stream, string)>();
+                    foreach (var item in itemsStream)
+                    {
+                        var targetPath = ResolveTargetPath(rootPath, item, artifact.Name);
+                        Stream stream = await this.DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, item, targetPath, cancellationToken);
+                        collection.Add((stream, targetPath));
+                    }
+                    return collection;
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount * 8,
+                    CancellationToken = cancellationToken,
+                });
+
+            var downloadBlock = NonSwallowingActionBlock.Create<(Stream stream, string targetPath)>(
+                item =>
+                {
+                    using (item.stream)
+                    {
+                        int retryCount = 0;
+                        try
+                        {
+                            using (var fileStream = new FileStream(item.targetPath, FileMode.Create))
+                            {
+                                item.stream.CopyTo(fileStream);
+                            }
+                        }
+                        catch (IOException exception) when (retryCount < 3)
+                        {
+                            Console.WriteLine($"Exception caught: {exception.Message}, on retry count {retryCount}, Retrying\n");
+                            retryCount++;
+                        }
+                    }               
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount,
+                    CancellationToken = cancellationToken,
```

Comment:
why is this different from the fetch block, I think they should have the same parallelism look at DedupManifestArtifactClient for reference.